### ISSUE TITLE
Why padding is needed?

### ISSUE THREAD

**Issue by pcpLiu** [otto-schnurr/MetalMatrixMultiply-swift#9]
Hi,

I'm studying the Metal programming and I checked the Apple example project and yours. Both do some padding for matrix C, which I was not very clear.

Why we need padding here?

Thanks!

**Comment by otto-schnurr** [comment]
Hello Zhonghao,

From what I understand, it's a tradeoff of more memory in exchange for a simpler computation kernel. By padding, we allow every thread to [assume full blocks of input to work on][grid].

The second paragraph of this Stack Overflow answer [here][performance] offers a lot more detail.

In terms of where to start, the best approach I have found is getting some kind of introductory book on [CUDA]. Even though Metal isn't CUDA and even if you don't have an Nvidia GPU on your Mac to program CUDA with, I've found CUDA material to be helpful in terms of starting from the beginning, explaining what the concepts are and what the motivations are behind them. Many of the hardware concepts like thread groups map directly from CUDA to Metal.

Hope that helps.
-Otto

[grid]: https://github.com/otto-schnurr/MetalMatrixMultiply-swift/blob/master/Source/Pipeline/MultiplyMatrices.metal#L47-L48
[performance]: http://stackoverflow.com/a/12539751/148076
[CUDA]: https://www.amazon.com/CUDA-Example-Introduction-General-Purpose-Programming/dp/0131387685/
